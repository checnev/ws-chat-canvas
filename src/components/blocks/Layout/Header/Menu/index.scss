@import '~material-icons/css/material-icons.scss';

.headerMenu {
  &__items > li:first-child{
    margin-left: -12px;
  }

  &__link {
    display: inline-block;
    color: rgb(153, 153, 153);
    text-decoration: none;
    padding: 12px;
    border-bottom: 2px solid transparent;
  }

  &__burger {
    @include material-icons();
    @include material-icon('menu');
    padding: 10px 0;

    &_open {
      @include material-icon('close');
    }

    @media screen and (min-width: 600px) {
      display: none;
    }
  }

  .menu__link_active {
    color: rgb(30, 167, 253);
    border-color: rgb(30, 167, 253);
  }

  @media screen and (max-width: 599px) {
    .menu__items {
      display: none;
    }

    .menu__item {
      display: block;
      > .headerMenu__link {
        display:block;
        border-bottom-width: 1px;
        border-color: rgba(0, 0, 0, .25);
      }
    }

    &_open {
      .menu__items {
        position: absolute;
        display: block;
        width: 100%;
        top: calc(100% + 1px);
        left: 0;
        z-index: 1;
        background: rgb(255, 255, 255);
        padding: 10px 0;
      }
    }
  }
}
