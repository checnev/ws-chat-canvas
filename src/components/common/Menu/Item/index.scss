.menu {
  &__link {
    padding: 8px 12px;
  }
}
