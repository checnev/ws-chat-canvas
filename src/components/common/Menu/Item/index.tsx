import React from 'react';
import cn from 'utils/cn';
import './index.scss';

class MenuItem extends React.Component {

  render() {
    const menu = cn('menu');
    return (
      <li
        {...this.props}
        className={menu('item')}
        role="menuitem"
      />
    );
  }
}

export default MenuItem;
