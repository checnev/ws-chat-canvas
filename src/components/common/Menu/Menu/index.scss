.menu {
  &__items {
    list-style: none;
    padding-left: 0;
    margin: 0;

    &_orientation_horizontal {
      > li {
        display: inline-block;
      }
    }
  }
}
