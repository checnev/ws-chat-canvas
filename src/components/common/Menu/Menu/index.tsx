import React from 'react';
import clsx from 'clsx';
import cn from 'utils/cn';
import './index.scss';

interface MenuProps {
  className?: string;
  orientation?: 'horizontal' | 'vertical';
}

class Menu extends React.Component<MenuProps> {
  static defaultProps: Pick<MenuProps, 'orientation'> = {
    orientation: 'vertical',
  };

  render() {
    const menu = cn('menu');
    return (
      <nav className={clsx(this.props.className, menu())}>
        <ul
          className={menu('items', { orientation: this.props.orientation })}
          role="menu"
        >
          {this.props.children}
        </ul>
      </nav>
    );
  }
}

export default Menu;
