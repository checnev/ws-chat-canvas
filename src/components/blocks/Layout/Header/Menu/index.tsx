import React from 'react';
import cn from 'utils/cn';
import clsx from 'clsx';
import { Menu, MenuItem } from 'components/common/Menu';
import { NavLink } from 'react-router-dom';
import './index.scss';

const menuItems = [
  { anchor: 'WS Chat & Canvas', to: '/', exact: true },
  { anchor: 'Chat', to: '/chat' },
  { anchor: 'Canvas', to: '/canvas' },
];

interface HeaderMenuProps {
  className?: string;
}

interface HeaderMenuState {
  isOpen: boolean;
}

class HeaderMenu extends React.Component<HeaderMenuProps, HeaderMenuState> {
  menuWrapperRef = React.createRef<HTMLDivElement>();

  state = {
    isOpen: false,
  };

  toggleMenu = () => {
    this.setState((state) => ({
      isOpen: !state.isOpen,
    }));
  };

  closeMenu = (event: MouseEvent) => {
    if (!this.menuWrapperRef.current!.contains(event.target as Node)) {
      this.setState({ isOpen: false });
    }
  }

  componentDidMount() {
    document.documentElement.addEventListener('click', this.closeMenu);
  }

  componentWillUnmount() {
    document.documentElement.removeEventListener('click', this.closeMenu);
  }

  render() {
    const menu = cn('headerMenu');
    return (
      <div className={menu()} ref={this.menuWrapperRef}>
        <div
          className={menu('burger', { open: this.state.isOpen })}
          onClick={this.toggleMenu}
        >
        </div>
        <Menu
          className={clsx(this.props.className, menu({ open: this.state.isOpen }))}
          orientation="horizontal"
        >
          {menuItems.map((menuItem) => (
            <MenuItem key={menuItem.to}>
              <NavLink
                to={menuItem.to}
                className={menu('link')}
                activeClassName="menu__link_active"
                exact={menuItem.exact}
              >
                {menuItem.anchor}
              </NavLink>
            </MenuItem>
          ))}
        </Menu>
      </div>
    );
  }
}

export default HeaderMenu;
