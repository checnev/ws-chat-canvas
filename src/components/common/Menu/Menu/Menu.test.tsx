import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import pretty from 'pretty';
import Menu from 'components/common/Menu/Menu';

let container: any = null;

beforeEach(() => {
  container = document.createElement('div');
  document.body.append(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it('should render Menu', () => {
  act(() => {
    render(<Menu></Menu>, container);
  });

  expect(
    pretty(container.innerHTML)
  ).toMatchSnapshot();
});
