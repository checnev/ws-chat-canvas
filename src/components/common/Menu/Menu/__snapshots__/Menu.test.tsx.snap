// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Menu 1`] = `
"<nav class=\\"menu\\">
  <ul class=\\"menu__items menu__items_orientation_vertical\\" role=\\"menu\\"></ul>
</nav>"
`;
